File: codesandbox/pkg/firecracker/vm.go
package firecracker

import (
	"context"
	"fmt"
	"io"
	"net"
	"os"
	"path/filepath"
	"time"

	firecracker "github.com/firecracker-microvm/firecracker-go-sdk"
	models "github.com/firecracker-microvm/firecracker-go-sdk/client/models"
	"github.com/google/uuid"
	log "github.com/sirupsen/logrus"
)

// VMConfig holds configuration for a Firecracker VM
type VMConfig struct {
	ID            string
	KernelPath    string
	RootfsPath    string
	MemSizeMib    int64
	CPUCount      int64
	NetworkConfig *NetworkConfig
}

// NetworkConfig holds network configuration for the VM
type NetworkConfig struct {
	HostDeviceName string
	IPAddr         string
	GatewayAddr    string
	Netmask        string
}

// VM represents a Firecracker microVM instance
type VM struct {
	ID      string
	Machine *firecracker.Machine
	Config  *VMConfig
	IPAddr  string
	ctx     context.Context
	cancel  context.CancelFunc
}

// Manager handles VM lifecycle
type Manager struct {
	vmPool       []*VM
	availableVMs chan *VM
	rootfsPath   string
	kernelPath   string
	config       *ManagerConfig
	netManager   *NetworkManager
}

// ManagerConfig holds configuration for the VM manager
type ManagerConfig struct {
	PoolSize      int
	RootfsPath    string
	KernelPath    string
	MemSizeMib    int64
	CPUCount      int64
	NetworkPrefix string // e.g., "172.16.0"
}

// NewManager creates a new VM manager
func NewManager(config *ManagerConfig) (*Manager, error) {
	// Create network manager
	netManager := NewNetworkManager(config.NetworkPrefix)
	if err := netManager.Setup(); err != nil {
		return nil, fmt.Errorf("failed to setup network: %w", err)
	}

	manager := &Manager{
		vmPool:       make([]*VM, 0, config.PoolSize),
		availableVMs: make(chan *VM, config.PoolSize),
		rootfsPath:   config.RootfsPath,
		kernelPath:   config.KernelPath,
		config:       config,
		netManager:   netManager,
	}

	// Initialize VM pool
	for i := 0; i < config.PoolSize; i++ {
		vm, err := manager.createVM(i)
		if err != nil {
			log.WithError(err).Error("Failed to create VM")
			continue
		}
		manager.vmPool = append(manager.vmPool, vm)
		manager.availableVMs <- vm
	}

	return manager, nil
}

// createVM creates a new Firecracker VM
func (m *Manager) createVM(index int) (*VM, error) {
	vmID := uuid.New().String()
	
	// Create VM-specific rootfs (copy-on-write)
	vmRootfs := filepath.Join("/tmp", fmt.Sprintf("rootfs-%s.ext4", vmID))
	if err := copyFile(m.rootfsPath, vmRootfs); err != nil {
		return nil, fmt.Errorf("failed to copy rootfs: %w", err)
	}

	// Network configuration
	netConfig := &NetworkConfig{
		HostDeviceName: fmt.Sprintf("fc-tap-%d", index),
		IPAddr:         fmt.Sprintf("%s.%d", m.config.NetworkPrefix, 100+index),
		GatewayAddr:    fmt.Sprintf("%s.1", m.config.NetworkPrefix),
		Netmask:        "255.255.255.0",
	}

	vmConfig := &VMConfig{
		ID:            vmID,
		KernelPath:    m.kernelPath,
		RootfsPath:    vmRootfs,
		MemSizeMib:    m.config.MemSizeMib,
		CPUCount:      m.config.CPUCount,
		NetworkConfig: netConfig,
	}

	return m.startVM(vmConfig, index)
}

// startVM starts a new Firecracker VM
func (m *Manager) startVM(config *VMConfig, index int) (*VM, error) {
	ctx, cancel := context.WithCancel(context.Background())
	
	// Don't pre-create TAP - Firecracker will create it
	// We'll add it to the bridge after VM starts

	// Socket path for Firecracker API
	socketPath := filepath.Join("/tmp", fmt.Sprintf("firecracker-%s.sock", config.ID))

	// Generate MAC address for the VM
	macAddr := generateMAC()

	// Pre-create TAP device with proper ownership
	if err := m.netManager.CreateTAPDevice(config.NetworkConfig.HostDeviceName); err != nil {
		cancel()
		return nil, fmt.Errorf("failed to create TAP device: %w", err)
	}
	
	// VM configuration
	cfg := firecracker.Config{
		SocketPath:      socketPath,
		KernelImagePath: config.KernelPath,
		Drives: []models.Drive{{
			DriveID:      firecracker.String("rootfs"),
			PathOnHost:   firecracker.String(config.RootfsPath),
			IsRootDevice: firecracker.Bool(true),
			IsReadOnly:   firecracker.Bool(false),
		}},
		MachineCfg: models.MachineConfiguration{
			VcpuCount:  firecracker.Int64(config.CPUCount),
			MemSizeMib: firecracker.Int64(config.MemSizeMib),
		},
		NetworkInterfaces: []firecracker.NetworkInterface{{
			StaticConfiguration: &firecracker.StaticNetworkConfiguration{
				HostDevName: config.NetworkConfig.HostDeviceName,
				MacAddress:  macAddr,
			},
		}},
		KernelArgs: fmt.Sprintf("console=ttyS0 reboot=k panic=1 pci=off vm_id=%d init=/init", index),
	}

	// Create machine - the API server should be run with sudo for network permissions
	machine, err := firecracker.NewMachine(ctx, cfg)
	if err != nil {
		cancel()
		return nil, fmt.Errorf("failed to create machine: %w", err)
	}

	// Start the VM
	if err := machine.Start(ctx); err != nil {
		cancel()
		return nil, fmt.Errorf("failed to start machine: %w", err)
	}

	vm := &VM{
		ID:      config.ID,
		Machine: machine,
		Config:  config,
		IPAddr:  config.NetworkConfig.IPAddr,
		ctx:     ctx,
		cancel:  cancel,
	}

	// Wait for VM to be ready
	if err := vm.waitForReady(); err != nil {
		vm.Stop()
		return nil, fmt.Errorf("VM failed to become ready: %w", err)
	}

	return vm, nil
}

// GetVM retrieves an available VM from the pool
func (m *Manager) GetVM(ctx context.Context) (*VM, error) {
	select {
	case vm := <-m.availableVMs:
		return vm, nil
	case <-ctx.Done():
		return nil, ctx.Err()
	}
}

// ReturnVM returns a VM to the pool
func (m *Manager) ReturnVM(vm *VM) {
	// Reset VM state
	if err := vm.Reset(); err != nil {
		log.WithError(err).Error("Failed to reset VM")
		// Create a new VM to replace the broken one
		newVM, err := m.createVM(len(m.vmPool))
		if err != nil {
			log.WithError(err).Error("Failed to create replacement VM")
			return
		}
		vm = newVM
	}
	
	m.availableVMs <- vm
}

// Cleanup shuts down all VMs
func (m *Manager) Cleanup() {
	close(m.availableVMs)
	for _, vm := range m.vmPool {
		vm.Stop()
		// Clean up TAP device
		if vm.Config != nil && vm.Config.NetworkConfig != nil {
			m.netManager.DeleteTAPDevice(vm.Config.NetworkConfig.HostDeviceName)
		}
	}
	// Clean up network
	m.netManager.Cleanup()
}

// Stop stops the VM
func (vm *VM) Stop() error {
	if vm.cancel != nil {
		vm.cancel()
	}
	
	if vm.Machine != nil {
		if err := vm.Machine.StopVMM(); err != nil {
			return fmt.Errorf("failed to stop VMM: %w", err)
		}
	}
	
	// Cleanup resources
	if vm.Config != nil {
		os.Remove(vm.Config.RootfsPath)
		// Note: TAP device cleanup handled by VM manager
	}
	
	return nil
}

// Reset resets the VM to a clean state
func (vm *VM) Reset() error {
	// For now, we'll just ensure the VM is still responsive
	// In production, you might want to snapshot/restore the rootfs
	return vm.waitForReady()
}

// waitForReady waits for the VM to be ready to accept commands
func (vm *VM) waitForReady() error {
	timeout := time.After(30 * time.Second)
	ticker := time.NewTicker(1 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-timeout:
			return fmt.Errorf("timeout waiting for VM to be ready")
		case <-ticker.C:
			// Try to connect to the agent inside the VM
			conn, err := net.Dial("tcp", fmt.Sprintf("%s:8080", vm.IPAddr))
			if err == nil {
				conn.Close()
				return nil
			}
		}
	}
}

// Helper functions

func copyFile(src, dst string) error {
	source, err := os.Open(src)
	if err != nil {
		return err
	}
	defer source.Close()

	destination, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer destination.Close()

	_, err = io.Copy(destination, source)
	return err
}

func generateMAC() string {
	mac := make([]byte, 6)
	mac[0] = 0x02 // Locally administered
	for i := 1; i < 6; i++ {
		mac[i] = byte(time.Now().UnixNano() >> (8 * i) & 0xff)
	}
	return fmt.Sprintf("%02x:%02x:%02x:%02x:%02x:%02x",
		mac[0], mac[1], mac[2], mac[3], mac[4], mac[5])
}

File: codesandbox-rust/src/vm.rs
use crate::{Result, CodeboxError, Config, VMConfig, NetworkConfig};
use crate::network::NetworkManager;
use std::fs;
use std::path::Path;
use std::process::Stdio;
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::net::TcpStream;
use tokio::sync::{mpsc, Mutex};
use tokio::time::timeout;
use tracing::{debug, error, info};
use uuid::Uuid;

#[derive(Debug)]
pub struct VM {
    pub id: String,
    pub config: VMConfig,
    pub ip_addr: String,
    process: Option<tokio::process::Child>,
}

pub struct Manager {
    available_vms: Arc<Mutex<mpsc::UnboundedReceiver<VM>>>,
    available_sender: mpsc::UnboundedSender<VM>,
    config: Config,
    net_manager: NetworkManager,
}

impl VM {
    pub async fn stop(&mut self) -> Result<()> {
        if let Some(mut process) = self.process.take() {
            // Attempt graceful shutdown first
            let _ = process.kill().await;
            let _ = process.wait().await;
        }

        // Cleanup resources
        if Path::new(&self.config.rootfs_path).exists() {
            let _ = fs::remove_file(&self.config.rootfs_path);
        }

        Ok(())
    }

    pub async fn reset(&self) -> Result<()> {
        // For now, we'll just ensure the VM is still responsive
        // In production, you might want to snapshot/restore the rootfs
        self.wait_for_ready().await
    }

    pub async fn wait_for_ready(&self) -> Result<()> {
        let timeout_duration = Duration::from_secs(30);
        let mut interval = tokio::time::interval(Duration::from_secs(1));
        
        let start = Instant::now();
        
        loop {
            if start.elapsed() > timeout_duration {
                return Err(CodeboxError::Timeout("timeout waiting for VM to be ready".to_string()));
            }
            
            interval.tick().await;
            
            // Try to connect to the agent inside the VM
            match timeout(Duration::from_secs(2), TcpStream::connect(format!("{}:8080", self.ip_addr))).await {
                Ok(Ok(_)) => {
                    debug!("VM {} is ready", self.id);
                    return Ok(());
                }
                Ok(Err(_)) | Err(_) => {
                    // Connection failed, continue waiting
                    continue;
                }
            }
        }
    }
}

impl Manager {
    pub async fn new(config: Config) -> Result<Self> {
        // Create network manager
        let net_manager = NetworkManager::new(config.network_prefix.clone());
        net_manager.setup()?;

        let (sender, receiver) = mpsc::unbounded_channel();

        let manager = Self {
            available_vms: Arc::new(Mutex::new(receiver)),
            available_sender: sender,
            config,
            net_manager,
        };

        // Initialize VM pool
        for i in 0..manager.config.vm_pool_size {
            match manager.create_vm(i).await {
                Ok(vm) => {
                    let _ = manager.available_sender.send(vm);
                }
                Err(e) => {
                    error!("Failed to create VM {}: {}", i, e);
                    continue;
                }
            }
        }

        info!("VM manager initialized with {} VMs", manager.config.vm_pool_size);
        Ok(manager)
    }

    async fn create_vm(&self, index: usize) -> Result<VM> {
        let vm_id = Uuid::new_v4().to_string();
        
        // Create VM-specific rootfs (copy-on-write)
        let vm_rootfs = format!("/tmp/rootfs-{}.ext4", vm_id);
        self.copy_file(&self.config.rootfs_path, &vm_rootfs)?;

        // Network configuration
        let net_config = NetworkConfig {
            host_device_name: format!("fc-tap-{}", index),
            ip_addr: format!("{}.{}", self.config.network_prefix, 100 + index),
            gateway_addr: format!("{}.1", self.config.network_prefix),
            netmask: "255.255.255.0".to_string(),
        };

        let vm_config = VMConfig {
            id: vm_id.clone(),
            kernel_path: self.config.kernel_path.clone(),
            rootfs_path: vm_rootfs,
            mem_size_mib: self.config.mem_size_mib,
            cpu_count: self.config.cpu_count,
            network_config: net_config,
        };

        self.start_vm(vm_config, index).await
    }

    async fn start_vm(&self, config: VMConfig, index: usize) -> Result<VM> {
        // Pre-create TAP device with proper ownership
        self.net_manager.create_tap_device(&config.network_config.host_device_name)?;

        // Socket path for Firecracker API
        let socket_path = format!("/tmp/firecracker-{}.sock", config.id);

        // Generate MAC address for the VM
        let mac_addr = self.generate_mac();

        // Create Firecracker config JSON
        let firecracker_config = serde_json::json!({
            "boot-source": {
                "kernel_image_path": config.kernel_path,
                "boot_args": format!("console=ttyS0 reboot=k panic=1 pci=off vm_id={} init=/init", index)
            },
            "drives": [{
                "drive_id": "rootfs",
                "path_on_host": config.rootfs_path,
                "is_root_device": true,
                "is_read_only": false
            }],
            "machine-config": {
                "vcpu_count": config.cpu_count,
                "mem_size_mib": config.mem_size_mib
            },
            "network-interfaces": [{
                "iface_id": "eth0",
                "guest_mac": mac_addr,
                "host_dev_name": config.network_config.host_device_name
            }]
        });

        // Write config to temporary file
        let config_path = format!("/tmp/firecracker-config-{}.json", config.id);
        fs::write(&config_path, firecracker_config.to_string())?;

        // Start Firecracker process
        let mut cmd = tokio::process::Command::new("firecracker");
        cmd.arg("--api-sock").arg(&socket_path)
            .arg("--config-file").arg(&config_path)
            .stdout(Stdio::piped())
            .stderr(Stdio::piped());

        let process = cmd.spawn()?;

        let ip_addr = config.network_config.ip_addr.clone();
        let vm_id = config.id.clone();
        
        let mut vm = VM {
            id: vm_id,
            config,
            ip_addr,
            process: Some(process),
        };

        // Wait for VM to be ready
        if let Err(e) = vm.wait_for_ready().await {
            vm.stop().await?;
            return Err(CodeboxError::Vm(format!("VM failed to become ready: {}", e)));
        }

        // Cleanup temporary config file
        let _ = fs::remove_file(&config_path);

        info!("VM {} started successfully", vm.id);
        Ok(vm)
    }

    pub async fn get_vm(&self) -> Result<VM> {
        let mut receiver = self.available_vms.lock().await;
        
        match timeout(Duration::from_secs(5), receiver.recv()).await {
            Ok(Some(vm)) => Ok(vm),
            Ok(None) => Err(CodeboxError::ResourceUnavailable("VM pool is closed".to_string())),
            Err(_) => Err(CodeboxError::ResourceUnavailable("No available VMs".to_string())),
        }
    }

    pub async fn return_vm(&self, mut vm: VM) {
        // Reset VM state
        match vm.reset().await {
            Ok(_) => {
                let _ = self.available_sender.send(vm);
            }
            Err(e) => {
                error!("Failed to reset VM {}: {}", vm.id, e);
                // Create a new VM to replace the broken one
                let _ = vm.stop().await;
                
                // Try to create a replacement VM
                match self.create_vm(0).await {
                    Ok(new_vm) => {
                        let _ = self.available_sender.send(new_vm);
                    }
                    Err(e) => {
                        error!("Failed to create replacement VM: {}", e);
                    }
                }
            }
        }
    }

    pub async fn cleanup(&mut self) {
        info!("Cleaning up VM manager");
        
        // Note: In the original Go version, VMs were tracked in vm_pool for cleanup
        // In this Rust version, VMs are managed through channels, so we can't easily
        // iterate over all VMs for cleanup. In production, you might want to maintain
        // a separate registry of VM IDs for cleanup purposes.
        
        // Clean up network
        if let Err(e) = self.net_manager.cleanup() {
            error!("Failed to cleanup network manager: {}", e);
        }
    }

    fn copy_file(&self, src: &str, dst: &str) -> Result<()> {
        fs::copy(src, dst)?;
        Ok(())
    }

    fn generate_mac(&self) -> String {
        use std::time::{SystemTime, UNIX_EPOCH};
        
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        
        let mac: [u8; 6] = [
            0x02, // Locally administered
            ((now >> 8) & 0xff) as u8,
            ((now >> 16) & 0xff) as u8,
            ((now >> 24) & 0xff) as u8,
            ((now >> 32) & 0xff) as u8,
            ((now >> 40) & 0xff) as u8,
        ];
        
        format!("{:02x}:{:02x}:{:02x}:{:02x}:{:02x}:{:02x}",
            mac[0], mac[1], mac[2], mac[3], mac[4], mac[5])
    }
}